flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/codegen/WatermarkGeneratorCodeGenTest.scala

```diff
@@ -0,0 +1,123 @@
+import org.apache.flink.table.planner.runtime.utils.JavaUserDefinedScalarFunctions.JavaFunc5
+import org.apache.flink.table.runtime.operators.wmassigners.WatermarkGenerator
+import org.apache.flink.table.types.logical.{IntType, TimestampType}
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Test
+
+import java.lang.{Integer => JInt, Long => JLong}
+
+/**
+  * Tests the generated [[WatermarkGenerator]] from [[WatermarkGeneratorCodeGenerator]].
+  */
+class WatermarkGeneratorCodeGenTest {
+
+  val tEnv: StreamTableEnvironmentImpl = StreamTableEnvironmentImpl.create(
+    StreamExecutionEnvironment.getExecutionEnvironment,
+    EnvironmentSettings.newInstance().useBlinkPlanner().build(),
+    new TableConfig)
+  val data = List(
+    GenericRow.of(JLong.valueOf(1000L), JInt.valueOf(5)),
+    GenericRow.of(null, JInt.valueOf(4)),
+    GenericRow.of(JLong.valueOf(3000L), null),
+    GenericRow.of(JLong.valueOf(5000L), JInt.valueOf(3)),
+    GenericRow.of(JLong.valueOf(4000L), JInt.valueOf(10)),
+    GenericRow.of(JLong.valueOf(6000L), JInt.valueOf(8))
+  )
+
+  @Test
+  def testAscendingWatermark(): Unit = {
+    val generator = generateWatermarkGenerator("ts - INTERVAL '0.001' SECOND")
+    val results = data.map(d => generator.currentWatermark(d))
+    val expected = List(
+      JLong.valueOf(999L),
+      null,
+      JLong.valueOf(2999),
+      JLong.valueOf(4999),
+      JLong.valueOf(3999),
+      JLong.valueOf(5999))
+    assertEquals(expected, results)
+  }
+
+  @Test
+  def testBoundedOutOfOrderWatermark(): Unit = {
+    val generator = generateWatermarkGenerator("ts - INTERVAL '5' SECOND")
+    val results = data.map(d => generator.currentWatermark(d))
+    val expected = List(
+      JLong.valueOf(-4000L),
+      null,
+      JLong.valueOf(-2000L),
+      JLong.valueOf(0L),
+      JLong.valueOf(-1000L),
+      JLong.valueOf(1000L))
+    assertEquals(expected, results)
+  }
+
+  @Test
+  def testCustomizedWatermark(): Unit = {
+    JavaFunc5.openCalled = false
+    JavaFunc5.closeCalled = false
+    tEnv.registerFunction("myFunc", new JavaFunc5)
+    val generator = generateWatermarkGenerator("myFunc(ts, `offset`)")
+    // mock open and close invoking
+    generator.setRuntimeContext(new MockStreamingRuntimeContext(false, 1, 1))
+    generator.open(new Configuration())
+    val results = data.map(d => generator.currentWatermark(d))
+    generator.close()
+    val expected = List(
+      JLong.valueOf(995L),
+      null,
+      null,
+      JLong.valueOf(4997L),
+      JLong.valueOf(3990L),
+      JLong.valueOf(5992L))
+    assertEquals(expected, results)
+    assertTrue(JavaFunc5.openCalled)
+    assertTrue(JavaFunc5.closeCalled)
+  }
+
+  private def generateWatermarkGenerator(expr: String): WatermarkGenerator = {
+    val planner = tEnv.getPlanner.asInstanceOf[StreamPlanner]
+    val tableRowType = planner.getTypeFactory.buildRelNodeRowType(
```

Comment:
We do not need to create a `TableEnvironment`, Mock a `FlinkPlannerImpl` already satisfies your needs.

Reply:
However, I need to register UDFs in the tests. And mocking `FlinkPlannerImpl` is not easy. 
I think creating a `TableEnvironment` is not much overhead, it will not spawn cluster.

Reply:
You can take SqlToOperationConverterTest as a reference.